Documentation for operator privileges (#68964)
elastic/elasticsearch: x-pack/docs/en/security/operator-privileges/configure-operator-privileges.asciidoc

[role="xpack"]
[testenv="enterprise"]
[[configure-operator-privileges]]
=== Configure operator privileges

==== Enable operator privileges

> Operator privileges need to be enabled explicitly on each node in the cluster.

Reviewer: What happens if a node doesn't enable operator privileges while others are doing?

Author: You will get inconsistent result depending on which node the request hit first and where the request gets executed. Assuming the user has sufficient RBAC privileges for an action:
1. If the request hits a node with operator privileges turn on, it should generally work correctly. We could have 2 scenarios:
    * if the user is not an operator, the request will be rejected on this node and never get passed to another node.
    * if the user is an operator, the request will be authorized by the local node. When the request gets passed to another node where operator privileges are disabled, it will still work because it just become a normal request.
2. If the request hits a node with operator privileges disabled, things could break. The request is always authorized locally. Now:
    * if the request can be executed locally, it will be and things break.
    * if the request needs to be passed to another node where operator privileges are enabled, the remote node most likely will reject the request so that things still kinda work. A theoretical possibility is that the child action sent to the remote node is different from the parent action and it may or may not be operator-only. We don't have such issue right now, but if it becomes the case in the future, things will get more inconsistent.

Reviewer: Thanks for the details. Would it make sense to explicitly warn that you can get an inconsistent behavior otherwise? Just to reinforce that.

Author: Added a warning about it. But didn't go into the details because it may be too distracting and the important message is just "please do not it".

Reviewer: Thanks, LGTM! 👍

[role="xpack"]
[testenv="enterprise"]
[[configure-operator-privileges]]
=== Configure operator privileges

==== Enable operator privileges

> Operator privileges need to be enabled explicitly on each node in the cluster.

Reviewer 2: ```suggestion
In order to use _Operator privileges_, they must be enabled explicitly on each node in the cluster.
```

This is done by configure the `elasticsearch.yml` file:

[source,yaml]
----------------------------
xpack.security.operator_privileges.enabled: true
----------------------------

If the node is already running before the above change is made, you need restart
it for the feature to take effect.

WARNING: The feature needs to be either enabled or disabled consistently across all nodes
in a cluster. Otherwise, you can get inconsistent behaviours depending on which node

> a request hits first and where it gets executed.

Reviewer 2: ```suggestion
first receives a request and which node executes it.
```

> 

Reviewer 2: I think there's a section missing here. You'll probably want to clean up the words a bit, but roughly:

> When operator privileges are enabled on a cluster, [specific functionality](https://elasticsearch_68964.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/operator-only-functionality.html) is restricted so that it may only be executed by those used who have been explicitly designated as operators.

That just makes the flow of thought a little more clear -  it walks the reader through the process of
- turn on the setting,
- what did that do?
- okay now how do I how do I designate which users are intended to be operators?

Author: Added two sentences to make the transition smoother as suggested.

==== Designate operator users

> Operator users are just normal {es} users with special privileges to perform

Reviewer 2: I think "privileges" here, makes people think of RBAC privileges, so it might be better to introduce a different word while we're still explaining the concept.
Overall, "Operator privileges" is fine, but for someone learning about it for the first time, they need to change their thinking, and reusing words to mean something new, before they're got their head around the idea itself, can be confusing.

```suggestion
Operator users are just normal {es} users with special rights to perform
```

operator-only functionalities. They are specified in an `operator_users.yml` file located
in `ES_PATH_CONF`. Similar to <<file-realm-configuration,other security config files>>,
the `operator_users.yml` file is local to a node and not globally to the cluster.
This means, in most cases, the same file should be distributed/copied to all nodes in
a cluster.

The `operator_users.yml` file defines a set of criteria that an authenticating
user must match to be considered as an operator.
The following snippet shows an example of such a file:

[source,yaml]
-----------------------------------
operator: <1>
  - usernames: ["system_agent_1","system_agent_2"] <2>
    realm_type: "file" <3>
    realm_name: "file1" <4>
    auth_type: "realm" <5>
-----------------------------------
<1> A fixed value of `operator` to signal the beginning of the definition.
<2> A list of usernames allowed for operator users. This field is mandatory.
<3> The type of the authenticating realm allowed for operator users. The default and only
    acceptable value is <<file-realm,`file`>>.

> <4> The name of the authenticating realm allowed for operator users. For the `file` realm, the name is ignored.

Reviewer 2: If the `realm_type` has to be file, and `realm_name` is ignored for file, doesn't that mean `realm_name` is always ignored, and we can drop it from the docs?

Reviewer 2: I think this question is still outstanding.

Author: Somehow I missed this one. `realm_name` is now removed.

<5> The authentication type allowed for operator users. The default and only acceptable value is `realm`.

You must specify at least the `usernames` field. Other fields will use their default values if not specified.
All fields must be matched for a user to be qualified as an operator. You can also specify multiple groups of
criteria. This is currently not very useful since this feature does not yet support other realms and
authentication types. Besides the configuration, there are also two implicit rules:

> 1. If the authenticating user <<run-as-privilege,run as>> another user, neither of them will be considered

Reviewer 2: ```suggestion
1. If the authenticating user <<run-as-privilege,runs as>> another user, neither of them will be considered
```

as an operator.
2. All <<internal-users>> are implicitly operator users.

> IMPORTANT: Once a user is designated as an operator, it still subjects to regular RBAC <<authorization>> checks.

Reviewer 2: ```suggestion
IMPORTANT: Once a user is designated as an operator, it is still subject to regular RBAC <<authorization>> checks.
```

> It is entirely possible that an operator user fails to perform certain actions due to RBAC check failures.
> In short, an operator is *not* automatically a `superuser`.

Reviewer 2: ```suggestion
That is, in addition to specifying that the user is an operator you must also grant them the necessary
{es} roles to perform the functions that you wish to use.
Consequently, it is entirely possible that an operator user may encounter an "access denied" error,
and fail to perform certain actions due to RBAC check failures.
In short, an operator is *not* automatically a `superuser`.
```
